### PR TITLE
BaiTap04 - DuongNT

### PR DESCRIPTION
**Pull request by at-duongnguyen**

### CODE REVIEW

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/AppDelegate.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+class AppDelegate: UIResponder, UIApplicationDelegate {
+
+    var window: UIWindow?
+
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        window = UIWindow(frame: UIScreen.main.bounds)
+        if let window = window {
+            window.backgroundColor = .white
+            window.makeKeyAndVisible()
+            //let viewController = ViewController()
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/AppDelegate.swift]: Remove unused code

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+//
+//  Created by Nguyen Duong on 6/27/19.
+//  Copyright © 2019 Nguyen Duong. All rights reserved.
+//
+
+import UIKit
+
+class ViewController: UIViewController {
+
+    @IBOutlet weak var chuoiView: UIView!
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift]: Add `private` weak for all outlets in your code! Please check it!

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setViewUI()
+    }
+    
+    func setViewUI() {
+        chuoiView.layer.cornerRadius = 10
+        chuoiView.backgroundColor = .blue
+        chuoiLabel.text = "Chuoi"
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift]: You can set text in UI, no need in code

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        taoLabel.textColor = .white
+        taoNumLabel.textColor = .white
+        taoLabel.textAlignment = .center
+        taoNumLabel.textAlignment = .center
+        btnAddAll.backgroundColor = .orange
+        btnAddAll.layer.cornerRadius = 10
+        btnAddAll.frame = CGRect(x: UIScreen.main.bounds.width / 2 - 50, y: UIScreen.main.bounds.height / 2, width: 100, height: 50)
+    }
+    
+    @IBAction func chuoiButtonTouchUpInside(_ sender: UIView) {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift]: Add `private` func for all Actions. Please check it in all your code!

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+class LoginViewController: UIViewController {
+
+    @IBOutlet weak var userNameTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var lblError: UILabel!
+    @IBOutlet weak var btnLogin: UIButton!
+    @IBOutlet weak var btnClear: UIButton!
+    
+    @IBAction func btnLogin(_ sender: UIButton) {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: Move action function after life cycle function.
Rename, should be `loginButtonTouchUpInside`. Please check it in all your code

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        lblError.textColor = .red
+    }
+    
+    private func configTextFields() {
+        userNameTextField.delegate = self
+        passwordTextField.delegate = self
+    }
+    
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+        self.view.endEditing(true)
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: No need self, call `super.touchesBegan` before

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+//  Copyright © 2019 Nguyen Duong. All rights reserved.
+//
+
+import UIKit
+
+class LoginViewController: UIViewController {
+
+    @IBOutlet weak var userNameTextField: UITextField!
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var lblError: UILabel!
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: Rename, should be `errorLabel`

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    @IBOutlet weak var btnLogin: UIButton!
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: Rename, should be `loginButton`. Please check it in all your code

File: BaiTap04/DuongNT/MyPlayground.playground/Contents.swift [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import UIKit
```
Comment (at-bienle) [inline BaiTap04/DuongNT/MyPlayground.playground/Contents.swift]: Why you have code of `BasicSwift` in pull request of `Outlet & Action`

Comment (at-duongnguyen) [reply BaiTap04/DuongNT/MyPlayground.playground/Contents.swift]: Oh, this file test result in code fight.

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    @IBOutlet private weak var chuoiView: UIView!
+    @IBOutlet private weak var chuoiLabel: UILabel!
+    @IBOutlet private weak var chuoiNumLabel: UILabel!
+    @IBOutlet private weak var nhoView: UIView!
+    @IBOutlet private weak var nhoNumLabel: UILabel!
+    @IBOutlet private weak var nhoLabel: UILabel!
+    @IBOutlet private weak var taoView: UIView!
+    @IBOutlet private weak var taoLabel: UILabel!
+    @IBOutlet private weak var taoNumLabel: UILabel!
+    @IBOutlet private weak var btnAddAll: UIButton!
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap2/ViewController.swift]: Rename

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap3/ListUsersViewController.swift [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            view.addSubview(views)
+        }
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        showListAvatar()
+    }
+    
+    @objc private func buttonDidClick(_ button: UIButton) {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap3/ListUsersViewController.swift]: Rename

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap3/ListUsersViewController.swift [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                y += 130 + 20
+            }
+            let frame =  CGRect(x: x, y: y, width: 100 , height: 130 )
+            x += 120
+            let views = setAvatar(frame: frame, name: users[index].name, avatarImage: users[index].avatar, tag: index)
+            view.addSubview(views)
+        }
+    }
+    
+    override func viewDidLoad() {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap3/ListUsersViewController.swift]: Move life cycle function below properties.
`Properties` -> `Life cycle function` -> `Basic override view function (touch move, touch began,...)` -> `Private/public custom function` -> `IBAction function`

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    @IBOutlet private weak var loginButton: UIButton!
+    @IBOutlet private weak var clearButton: UIButton!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setButton()
+        configTextFields()
+    }
+
+    @IBAction private func loginButtonTouchUpInside(_ sender: UIButton) {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: `Properties` -> `Life cycle function` -> `Basic override view function (touch move, touch began,...)` -> `Private/public custom function` -> `IBAction function`

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap5/SliderViewController.swift [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+//
+//  Created by Nguyen Duong on 7/1/19.
+//  Copyright © 2019 Nguyen Duong. All rights reserved.
+//
+
+import UIKit
+
+class SliderViewController: UIViewController {
+
+    @IBOutlet weak var lblTitle: UILabel!
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap5/SliderViewController.swift]: Please check all code in this class

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    @IBOutlet private weak var clearButton: UIButton!
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setButton()
+        configTextFields()
+    }
+    
+    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+        self.view.endEditing(true)
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap4/LoginViewController.swift]: No need `self`, call `super.touchesBegan()` first

File: BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap5/SliderViewController.swift [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        colorChangeView(red: red.value, green: green.value, blue: blue.value)
+        titleLabel.text = "Color (R: \(Int(red.value)), G: \(Int(green.value)), B: \(Int(blue.value)))"
+        titleLabel.font = UIFont.boldSystemFont(ofSize: 16)
+    }
+    
+    private func colorChangeView(red: Float, green: Float, blue: Float) {
+        mainView.backgroundColor = UIColor(red: CGFloat(red / 255), green: CGFloat(green / 255), blue: CGFloat(blue / 255), alpha: 1)
+    }
+    
+    @IBAction func changeSlider(_ sender: UISlider) {
```
Comment (at-bienle) [inline BaiTap04/DuongNT/BaiTap2/BaiTap2/BaiTap5/SliderViewController.swift]: Use `private`, name function is correct?

**Pull request merged**